SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,625 @@
+
+    callSites.forEach { callSite in
+        log("PartialApply call site: \(callSite.applySite)")
+        log("Passed in closures: ")
+        for index in callSite.closureArgDescriptors.indices {
+            var closureArgDescriptor = callSite.closureArgDescriptors[index]
+            log("\(index+1). \(closureArgDescriptor.closureInfo.closure)")
+        }
+    }
+    log("\n")
+}
+
+// =========== AutodiffClosureSpecializer ========== //
+
+fileprivate extension Dictionary<PartialApplyInst, CallSiteDescriptor> {
+    func intoCallSites(_ context: FunctionPassContext) -> Stack<CallSiteDescriptor> {
+        var callSites = Stack<CallSiteDescriptor>(context)
+        callSites.append(contentsOf: self.values)
+        return callSites
+    }    
+}
+
+struct AutodiffClosureSpecializer: ClosureSpecializer {
+    typealias ApplySiteType = ApplySite
+
+    let specializationLevelLimit = 2
+    let context: FunctionPassContext
+
+    init(_ context: FunctionPassContext) {
+        self.context = context
+    }
+
+    /// If the closure obtained from this partial_apply is not the pullback
+    /// returned from this autodiff VJP, we have nothing to do.
+    private func isPullbackInResultOfAutodiffVJP(pai: PartialApplyInst) -> Bool {
+        if let use = pai.result.firstUse {
+            if let tupleInst = use.instruction as? TupleInst {
+                let maybeReturnedTuple = tupleInst.result
+                let vjp = pai.parentFunction
+                if let returnInst = vjp.returnInstruction {
+                    if maybeReturnedTuple == returnInst.returnedValue {
+                        return true
+                    }
+                }
+            }
+        }
+
+        return false
+    }
+
+    func gatherCallSites(_ caller: Function) -> Stack<CallSiteDescriptor> {
+        // We should not look at reabstraction closures twice who we
+        // ultimately ended up using as an argument that we specialize on.
+        var usedReabstractionClosures = InstructionSet(self.context)
+        defer {
+            usedReabstractionClosures.deinitialize()
+        }
+
+        var paiToCallSiteDesc: [PartialApplyInst: CallSiteDescriptor] = [:]
+
+        for bb in caller.blocks {
+            for inst in bb.instructions {
+                // If this is not a closure that we support specializing, skip it.
+                if !isSupportedClosure(inst) {
+                    continue
+                }
+                let closureInst = inst as! SingleValueInstruction
+
+                // If this is a reabstraction thunk then we have already looked at it, skip it.
+                if usedReabstractionClosures.contains(closureInst) {
+                    continue
+                }
+                
+                var closureInfo: ClosureInfo? = nil
+
+                // Worklist of operands
+                var uses = Array(closureInst.uses)
+
+                // Live range end points.
+                var usePoints: [Instruction] = []
```

Comment:
Why not call it `var lifeRangeEndPoints`? Then you don't need the comment

Reply:
Changed